heat/core/indexing.py

```diff
@@ -56,39 +56,45 @@ def nonzero(x: DNDarray) -> DNDarray:
     except AttributeError:
         raise TypeError("Input must be a DNDarray, is {}".format(type(x)))
 
+    lcl_nonzero = torch.nonzero(input=local_x, as_tuple=False)
+
     if x.split is None:
-        # if there is no split then just return the values from torch
-        lcl_nonzero = torch.nonzero(input=local_x, as_tuple=False)
+        # if there is no split then just return the transpose of values from torch
+        lcl_nonzero = lcl_nonzero.transpose(0, 1)
+
         gout = list(lcl_nonzero.size())
         is_split = None
     else:
         # a is split
-        lcl_nonzero = torch.nonzero(input=local_x, as_tuple=False)
         # adjust local indices along split dimension
         _, displs = x.counts_displs()
         lcl_nonzero[..., x.split] += displs[x.comm.rank]
         del displs
+
+        lcl_nonzero = lcl_nonzero.transpose(0, 1)
+
         # get global size of split dimension
         gout = list(lcl_nonzero.size())
         gout[0] = x.comm.allreduce(gout[0], MPI.SUM)
```

Comment:
The original shape of `torch.nonzero` is _(number of nonzero elements, number of x dimensions)_

What the allreduce call does is sum all number of nonzero elements across all processes so that every process knows what the total number is (in order to set the global shape of the output DNDarray).

However, right now the allreduce call doesn't do what was intended anymore... `lcl_nonzero` has been transposed.